src/cluster_legacy.c

```diff
@@ -6183,41 +6430,128 @@ int clusterCommandSpecial(client *c) {
                                      CLUSTER_TODO_FSYNC_CONFIG);
             }
 
-            /* If this node was importing this slot, assigning the slot to
-             * itself also clears the importing status. */
-            if (n == myself &&
-                server.cluster->importing_slots_from[slot]) {
-                /* This slot was manually migrated, set this node configEpoch
-                 * to a new epoch so that the new version can be propagated
-                 * by the cluster.
-                 *
-                 * Note that if this ever results in a collision with another
-                 * node getting the same configEpoch, for example because a
-                 * failover happens at the same time we close the slot, the
-                 * configEpoch collision resolution will fix it assigning
-                 * a different epoch to each node. */
-                if (clusterBumpConfigEpochWithoutConsensus() == C_OK) {
-                    serverLog(LL_NOTICE,
-                        "configEpoch updated after importing slot %d", slot);
-                }
+            /* If this node or this node's primary was importing this slot,
+            * assigning the slot to itself also clears the importing status. */
+            if ((n == myself || n == myself->slaveof) &&
+                server.cluster->importing_slots_from[slot])
+            {
                 server.cluster->importing_slots_from[slot] = NULL;
-                /* After importing this slot, let the other nodes know as
-                 * soon as possible. */
-                clusterBroadcastPong(CLUSTER_BROADCAST_ALL);
+
+                /* Only primary broadcasts the updates */
+                if (n == myself) {
+                    /* This slot was manually migrated, set this node configEpoch
+                     * to a new epoch so that the new version can be propagated
+                     * by the cluster.
+                     *
+                     * Note that if this ever results in a collision with another
+                     * node getting the same configEpoch, for example because a
+                     * failover happens at the same time we close the slot, the
+                     * configEpoch collision resolution will fix it assigning
+                     * a different epoch to each node. */
+                    if (clusterBumpConfigEpochWithoutConsensus() == C_OK) {
+                        serverLog(LL_NOTICE,
+                                  "ConfigEpoch updated after importing slot %d",
+                                  slot);
+                    }
+                    /* After importing this slot, let the other nodes know as
+                     * soon as possible. */
+                    clusterBroadcastPong(CLUSTER_BROADCAST_ALL);
+                }
             }
+        } else if (!strcasecmp(c->argv[3]->ptr,"node") &&
+                  c->argc == 6 &&
+                !strcasecmp(c->argv[5]->ptr,"replicaonly"))
```

Comment:
I would like to revisit the addition of this "replicaonly" flag as well with the new core team. The primary rationale for going with this "replicaonly" flag at the time was because there was no precedent in a server initiated wait. I was not convinced by this argument. I think this flag is a band-aid solution and exposes the internal details unnecessarily. Public interfaces are not the right place for hacks and compromises. 

On the other hand, we could've implemented the HA contract of this slot ownership finalization logic completely in the server **without** any change on the client side and with a change that I think is quite scoped. 

How about I revise this PR with a server side implementation for comparison? 

@valkey-io/core-team

Reply:
I prefer server-initiated wait, i.e. blocking SETSLOT.

It's an admin command and as you say, it's better to expose the internals and avoid client changes.

Reply:
>It's an admin command and as you say, it's better to expose the internals and avoid client changes.

I assume you meant **not** expose? 

Yep. Let me work on a sever-initiated wait solution. I think I can come up with a cleaner solution than what I had previously.